File: src/HolzShots.Legacy/Drawing/Tools/Ellipse.vb
Imports System.Drawing.Drawing2D
Imports System.Numerics
Imports HolzShots.Drawing.Tools.UI

Namespace Drawing.Tools
    Friend Class Ellipse
        Implements ITool(Of EllipseSettings)
        Public Property BeginCoordinates As Vector2 Implements ITool(Of EllipseSettings).BeginCoordinates
        Public Property EndCoordinates As Vector2 Implements ITool(Of EllipseSettings).EndCoordinates

        Private Shared ReadOnly CursorInstance As New Cursor(My.Resources.crossMedium.GetHicon())
        Public ReadOnly Property Cursor As Cursor = CursorInstance Implements ITool(Of EllipseSettings).Cursor
        Public ReadOnly Property ToolType As ShotEditorTool = ShotEditorTool.Ellipse Implements ITool(Of EllipseSettings).ToolType

        Public ReadOnly Property SettingsControl As ISettingsControl(Of EllipseSettings) Implements ITool(Of EllipseSettings).SettingsControl

        Sub New()
            SettingsControl = New EllipseSettingsControl(EllipseSettings.Default)
        End Sub

        Public Sub LoadInitialSettings() Implements ITool(Of EllipseSettings).LoadInitialSettings
            If My.Settings.EllipseWidth > EllipseSettings.MaximumWidth OrElse My.Settings.EllipseWidth < EllipseSettings.MinimumWidth Then
                My.Settings.EllipseWidth = EllipseSettings.Default.Width
                My.Settings.Save()
            End If

            With SettingsControl.Settings
                .Color = My.Settings.EllipseColor
                .Width = My.Settings.EllipseWidth
                .Mode = If(My.Settings.UseBoxInsteadOfCirlce, EllipseMode.Rectangle, EllipseMode.Ellipse)
            End With
        End Sub

        Public Sub PersistSettings() Implements ITool(Of EllipseSettings).PersistSettings
            With SettingsControl.Settings
                .Color = My.Settings.EllipseColor
                .Width = My.Settings.EllipseWidth
                My.Settings.UseBoxInsteadOfCirlce = .Mode = EllipseMode.Rectangle
            End With
        End Sub

        Private Shared Function CreatePen(settings As EllipseSettings) As Pen
            Return New Pen(settings.Color, settings.Width) With {
                .DashStyle = DashStyle.Solid
            }
        End Function

        Public Sub RenderFinalImage(ByRef rawImage As Image) Implements ITool(Of EllipseSettings).RenderFinalImage
            Dim settings = SettingsControl.Settings

            Dim rect = Rectangle.Round(New RectangleF(
                Math.Min(EndCoordinates.X, BeginCoordinates.X),
                Math.Min(EndCoordinates.Y, BeginCoordinates.Y),
                Math.Max(Math.Abs(BeginCoordinates.X - EndCoordinates.X), settings.Width),
                Math.Max(Math.Abs(BeginCoordinates.Y - EndCoordinates.Y), settings.Width)
            ))

            Using g = Graphics.FromImage(rawImage)
                Using pen = CreatePen(settings)
                    If settings.Mode = EllipseMode.Rectangle Then
                        g.DrawRectangle(pen, rect)
                    Else
                        g.SmoothingMode = SmoothingMode.AntiAlias
                        g.DrawEllipse(pen, rect)
                    End If
                End Using
            End Using
        End Sub

        Public Sub RenderPreview(rawImage As Image, g As Graphics) Implements ITool(Of EllipseSettings).RenderPreview
            Dim settings = SettingsControl.Settings

            Dim rect = Rectangle.Round(New RectangleF(
                Math.Min(EndCoordinates.X, BeginCoordinates.X),
                Math.Min(EndCoordinates.Y, BeginCoordinates.Y),
                Math.Max(Math.Abs(BeginCoordinates.X - EndCoordinates.X), settings.Width),
                Math.Max(Math.Abs(BeginCoordinates.Y - EndCoordinates.Y), settings.Width)
            ))

            Using pen = CreatePen(settings)
                If settings.Mode = EllipseMode.Rectangle Then
                    g.DrawRectangle(pen, rect)
                Else
                    g.SmoothingMode = SmoothingMode.AntiAlias
                    g.DrawEllipse(pen, rect)
                End If
            End Using
        End Sub

        Public Sub MouseOnlyMoved(rawImage As Image, ByRef currentCursor As Cursor, e As MouseEventArgs) Implements ITool(Of EllipseSettings).MouseOnlyMoved
            ' Nothing to do here
        End Sub
        Public Sub MouseClicked(rawImage As Image, e As Vector2, ByRef currentCursor As Cursor, trigger As Control) Implements ITool(Of EllipseSettings).MouseClicked
            ' Nothing to do here
        End Sub
        Public Sub Dispose() Implements ITool(Of EllipseSettings).Dispose
        End Sub
    End Class
End Namespace

File: src/HolzShots.LegacyNew/Drawing/Tools/Ellipse.cs
using System.Drawing.Drawing2D;
using System.Numerics;
using HolzShots.Drawing.Tools.UI;

namespace HolzShots.Drawing.Tools;

public class Ellipse : ITool<EllipseSettings>
{
    public Vector2 BeginCoordinates { get; set; }
    public Vector2 EndCoordinates { get; set; }

    private static readonly Cursor _cursor = new(Properties.Resources.crossMedium.GetHicon());
    public Cursor Cursor { get; } = _cursor;
    public ShotEditorTool ToolType { get; } = ShotEditorTool.Ellipse;

    public ISettingsControl<EllipseSettings> SettingsControl { get; }

    public Ellipse()
    {
        SettingsControl = new EllipseSettingsControl(EllipseSettings.Default);
    }

    public void LoadInitialSettings()
    {
        if (Properties.Settings.Default.EllipseWidth > EllipseSettings.MaximumWidth || Properties.Settings.Default.EllipseWidth < EllipseSettings.MinimumWidth)
        {
            Properties.Settings.Default.EllipseWidth = EllipseSettings.Default.Width;
            Properties.Settings.Default.Save();
        }

        var settings = SettingsControl.Settings;
        settings.Color = Properties.Settings.Default.EllipseColor;
        settings.Width = Properties.Settings.Default.EllipseWidth;
        settings.Mode = (EllipseMode)Properties.Settings.Default.EllipseMode;
    }

    public void PersistSettings()
    {
        var settings = SettingsControl.Settings;
        settings.Color = Properties.Settings.Default.EllipseColor;
        settings.Width = Properties.Settings.Default.EllipseWidth;
        Properties.Settings.Default.EllipseMode = (int)settings.Mode;
    }

    private static Pen CreatePen(EllipseSettings settings)
    {
        return new Pen(settings.Color, settings.Width)
        {
            DashStyle = DashStyle.Solid
        };
    }

    public void RenderFinalImage(ref Image rawImage)
    {
        var settings = SettingsControl.Settings;

        var rect = Rectangle.Round(new RectangleF(
            Math.Min(EndCoordinates.X, BeginCoordinates.X),
            Math.Min(EndCoordinates.Y, BeginCoordinates.Y),
            Math.Max(Math.Abs(BeginCoordinates.X - EndCoordinates.X), settings.Width),
            Math.Max(Math.Abs(BeginCoordinates.Y - EndCoordinates.Y), settings.Width)
        ));

        using var g = Graphics.FromImage(rawImage);
        using var pen = CreatePen(settings);

        if (settings.Mode == EllipseMode.Rectangle)
        {
            g.DrawRectangle(pen, rect);
        }
        else
        {
            g.SmoothingMode = SmoothingMode.AntiAlias;
            g.DrawEllipse(pen, rect);
        }
    }

    public void RenderPreview(Image rawImage, Graphics g)
    {
        var settings = SettingsControl.Settings;

        var rect = Rectangle.Round(new RectangleF(
            Math.Min(EndCoordinates.X, BeginCoordinates.X),
            Math.Min(EndCoordinates.Y, BeginCoordinates.Y),
            Math.Max(Math.Abs(BeginCoordinates.X - EndCoordinates.X), settings.Width),
            Math.Max(Math.Abs(BeginCoordinates.Y - EndCoordinates.Y), settings.Width)
        ));

        using var pen = CreatePen(settings);
        if (settings.Mode == EllipseMode.Rectangle)
        {
            g.DrawRectangle(pen, rect);
        }
        else
        {
            g.SmoothingMode = SmoothingMode.AntiAlias;
            g.DrawEllipse(pen, rect);
        }
    }

    public void MouseOnlyMoved(Image rawImage, ref Cursor currentCursor, MouseEventArgs e) { }
    public void MouseClicked(Image rawImage, Vector2 e, ref Cursor currentCursor, Control trigger) { }
    public void Dispose() { }
}
